### PR TITLE
Cfq dropout

### PR DESCRIPTION
**Pull request by Ranapop**
This PR contains several changes regarding dropout:

* the vertical dropout no longer reuses the same mask across time steps
* the vertical and horizontal dropout are only applied during training
* vertical and horizontal dropout can be configured separately
* embedding dropout can be configured (as in Joey decoder)
* attention layer dropout can be configured (as in Joey decoder)

This changes were tested on cfq after the luong decoder replaced bahdanau's decoder. Before gradient clipping was added, the dropout made the model suddenly drop in accuracy, but with gradient clipping that issue was resolved. With these changes the loss and dev accuracy are also better correlated.

### CODE REVIEW

File: examples/cfq/models.py [hunk 1]
```diff
@@ -103,24 +107,22 @@ class MultilayerLSTM(nn.Module):
   def apply(self,
             num_layers: int,
             horizontal_dropout_masks: jnp.array,
-            vertical_dropout_masks: jnp.array,
-            input: jnp.array, previous_states: List):
+            dropout_rate: float,
```
Comment (marcvanzee) [inline examples/cfq/models.py]: Why don't you name this `vertical_dropout_rate`?

**Review by marcvanzee — APPROVED** [review]
Nice, `nn.dropout` is very useful! 😄

File: examples/cfq/models.py (hunk 1 above)
Comment (Ranapop) [reply examples/cfq/models.py]: I was trying to be consistent with the names from recurren.py but probably I should either change them everywhere or not. Will modify.
